### PR TITLE
Fix coffin burial

### LINKED ISSUE #54525: Can't bury sealed coffin
### Describe the bug

My mate died, so I built a coffin. Put him in it and then sealed it.  Attempted to bury it via the construction menu, but no matter where I tell it to be buried, it says "you cannot build there"; yet the construction menu doesn't specify what else is needed (such as a deep pit) that would stop me from using my chosen space for a grave.

It does specify 2 planks OR 20 stones, so I grabbed both just in case it wasn't OR but AND, but that doesn't matter.

### Steps to reproduce

1. make a coffin
2. seal it
3. attempt to bury it
4. notice you can't, even in simple grass/dirt areas.

### Expected behavior

Put my mate to rest

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 10.0 2009
- Game Version: 1e91f34 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Aftershock [aftershock],
    Blaze Industries [blazeindustries],
    C.R.I.T Expansion Mod [crt_expansion],
    Generic Guns [generic_guns],
    Alternative Map Key [alt_map_key],
    Crazy Cataclysm [crazy_cataclysm],
    Mining Mod [Mining_Mod],
    More Survivor Stuff [more_survivor_stuff],
    No Rail Stations [No_Rail_Stations],
    Cars to Wrecks [cars_to_wrecks],
    Graphical Overmap [Graphical_Overmap],
    Bionic Slots [cbm_slots],
    Disable NPC Needs [no_npc_food],
    Mutation Changes [Mutation_Changes],
    No Fungal Growth [no_fungal_growth],
    SpeedyDex [speedydex],
    Stats Through Kills [stats_through_kills],
    Stats Through Skills [StatsThroughSkills]
]

### Additional context

I also tried to build a shallow and deep pit next to it, but those spots are the same "you cannot build here"

### PR DESCRIPTION
**Pull request by dseguin**
#### Summary
None

#### Purpose of change
- Fixes #54525

#54234 caused some terrain constructions to fail if furniture exists on top, because the pre-construction flags will apply to the furniture instead of the terrain. This is intended to prevent furniture on simple-to-deconstruct terrain from being simple-deconstructed.

#### Describe the solution
Added a field to force pre-construction flags to apply to terrain:

```json
"pre_flags": [ "FLAT", { "flag": "DIGGABLE", "force_terrain": true } ]
```

With `"force_terrain"` set to true, the construction will check if the terrain has the flag instead of the furniture.

#### Describe alternatives you've considered
Adding a special case for burials in `can_construct()`, reverting #54234, etc.

#### Testing

Sealed coffin on diggable terrain:

coffin_out

After burying the sealed coffin via the construction menu:

coffin_in
msg_log

### CODE REVIEW

File: data/json/construction.json [hunk 1]
```diff
@@ -1951,7 +1951,7 @@
     "time": "60 m",
     "qualities": [ [ { "id": "HAMMER", "level": 1 } ], [ { "id": "DIG", "level": 2 } ] ],
     "components": [ [ [ "rock", 20 ], [ "2x4", 2 ] ] ],
-    "pre_flags": "DIGGABLE",
+    "pre_flags": { "flag": "DIGGABLE", "force_ter": true },
```
Comment (ZhilkinSerg) [inline data/json/construction.json:1954]: Why not to use full word, i.e. `force_terrain`?

Comment (dseguin) [reply data/json/construction.json:1954]: Sorry, I'm used to using shorthand variables.

Comment (dseguin) [reply data/json/construction.json:1954]:
```suggestion
    "pre_flags": { "flag": "DIGGABLE", "force_terrain": true },
```

File: doc/JSON_INFO.md [hunk 2]
```diff
@@ -1683,7 +1683,7 @@ request](https://github.com/CleverRaven/Cataclysm-DDA/pull/36657) and the
 "components": [ [ [ "spear_wood", 4 ], [ "pointy_stick", 4 ] ] ],   // Items used in construction
 "pre_special": "check_empty",                                       // Required something that isn't terrain
 "pre_terrain": "t_pit",                                             // Alternative to pre_special; Required terrain to build on
-"pre_flags": [ "WALL" ],                                            // Flags beginning terrain must have
+"pre_flags": [ "WALL", { "flag": "DIGGABLE", "force_ter": true } ], // Flags beginning furniture/terrain must have. force_ter forces the flag to apply to the underlying terrain
```
Comment (dseguin) [inline doc/JSON_INFO.md:1686]:
```suggestion
"pre_flags": [ "WALL", { "flag": "DIGGABLE", "force_terrain": true } ], // Flags beginning furniture/terrain must have. force_ter forces the flag to apply to the underlying terrain
```

File: src/construction.cpp [hunk 3]
```diff
@@ -1815,7 +1817,24 @@ void load_construction( const JsonObject &jo )
...
     }
 
-    con.pre_flags = jo.get_tags( "pre_flags" );
+    if( jo.has_member( "pre_flags" ) ) {
+        con.pre_flags.clear();
+        if( jo.has_string( "pre_flags" ) ) {
+            con.pre_flags.emplace( jo.get_string( "pre_flags" ), false );
+        } else if( jo.has_object( "pre_flags" ) ) {
+            JsonObject jflag = jo.get_object( "pre_flags" );
+            con.pre_flags.emplace( jflag.get_string( "flag" ), jflag.get_bool( "force_ter" ) );
```
Comment (dseguin) [inline src/construction.cpp:1826]:
```suggestion
            con.pre_flags.emplace( jflag.get_string( "flag" ), jflag.get_bool( "force_terrain" ) );
```

File: src/construction.cpp [hunk 4]
```diff
@@ -1815,7 +1817,24 @@ void load_construction( const JsonObject &jo )
...
+        } else if( jo.has_array( "pre_flags" ) ) {
+            for( JsonValue jval : jo.get_array( "pre_flags" ) ) {
+                if( jval.test_string() ) {
+                    con.pre_flags.emplace( jval.get_string(), false );
+                } else if( jval.test_object() ) {
+                    JsonObject jflag = jval.get_object();
+                    con.pre_flags.emplace( jflag.get_string( "flag" ), jflag.get_bool( "force_ter" ) );
```
Comment (dseguin) [inline src/construction.cpp:1833]:
```suggestion
                    con.pre_flags.emplace( jflag.get_string( "flag" ), jflag.get_bool( "force_terrain" ) );
```

**Pull request merged**